File: solutions/032/032.py
"""
Project Euler - Problem Solution 032
Problem Title - Pandigital products
Copyright (c) Justin McGettigan. All rights reserved.
https://github.com/jwmcgettigan/project-euler-solutions
"""

def is_pandigital(identity):
  return len(identity) == 9 and set(identity) == set('123456789')

def product_identity(multiplicand, multiplier):
  product = multiplicand*multiplier
  identity = '{}{}{}'.format(multiplicand, multiplier, product)
  return product, identity

def pandigital_products():
  products = set()

  for multiplicand in range(10):
    for multiplier in range(1000, 10000):
      product, identity = product_identity(multiplicand, multiplier)
      if len(identity) > 9: break
      if is_pandigital(identity):
        products.add(product)
  
  for multiplicand in range(10, 100):
    for multiplier in range(100, 1000):
      product, identity = product_identity(multiplicand, multiplier)
      if len(identity) > 9: break
      if is_pandigital(identity):
        products.add(product)
  
  return sum(products)

if __name__ == "__main__":
  print(pandigital_products())

File: solutions/032/032.js
/**
 * Project Euler - Problem Solution 032
 * Problem Title - Pandigital products
 * Copyright (c) Justin McGettigan. All rights reserved.
 * https://github.com/jwmcgettigan/project-euler-solutions
 */

/** Return the sum of all elements in an iterable of numbers. */
const sum = (iterable) => {
  let s = 0;
  iterable.forEach((num) => { s += num; });
  return s;
}

/** Turns iterable into a Set(). */
const set = (iterable) => new Set(iterable);

/** Compares whether two sets contain the same elements. */
function equalSets(a, b) {
  if (a.size !== b.size) return false;
  for (const i of a) if (!b.has(i)) return false;
  return true;
}

/** Checks if a number contains all digits 1-9 exactly once. */
const isPandigital = (identity) => (
  identity.length === 9 && equalSets(set(identity), set('123456789'))
);

const productIdentity = (multiplicand, multiplier) => {
  const product = multiplicand*multiplier;
  const identity = `${multiplicand}${multiplier}${product}`;
  return [product, identity];
}

const pandigitalProducts = () => {
  const products = set();
  for(let multiplicand = 0; multiplicand < 10; multiplicand++) {
    for(let multiplier = 1000; multiplier < 10000; multiplier++) {
      [product, identity] = productIdentity(multiplicand, multiplier);
      if(identity.length > 9) { break; }
      if(isPandigital(identity)) {
        products.add(product);
      }
    }
  }
  for(let multiplicand = 10; multiplicand < 100; multiplicand++) {
    for(let multiplier = 100; multiplier < 1000; multiplier++) {
      [product, identity] = productIdentity(multiplicand, multiplier);
      if(identity.length > 9) { break; }
      if(isPandigital(identity)) {
        products.add(product);
      }
    }
  }

  return sum(products);
};

console.log(pandigitalProducts());